modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java

```diff
@@ -3859,18 +3866,52 @@ public ImageWorker affine(AffineTransform tx, Interpolation interpolation, doubl
                         PlanarImage sourceImage = op.getSourceImage(0);
                         final ParameterBlock paramBlk = new ParameterBlock().addSource(sourceImage);
                         Object property = sourceImage.getProperty("ROI");
-                        if ((property == null) || property.equals(java.awt.Image.UndefinedProperty)
-                                || !(property instanceof ROI)) {
+                        if (((property == null) || property.equals(java.awt.Image.UndefinedProperty)
+                                || !(property instanceof ROI))) {
                             paramBlk.add(warp).add(interpolation).add(bgValues);
                             paramBlk.set(nodata, 4);
-                            if(isNoDataNeeded() && bgValues != null && bgValues.length > 0){
+                            // Try to reproject ROI after Warp
+                            ROI newROI = null;
+                            if(roi != null){
+                                ROI reprojectedROI = roi;
+                                try{
+                                    MathTransform inverse = originalTransform.inverse();
+                                    if(inverse instanceof AffineTransform){
+                                        AffineTransform inv = (AffineTransform)inverse;
+                                        newROI = reprojectedROI.transform(inv);
+                                    }
+                                }catch(Exception e){
+                                    if(LOGGER.isLoggable(Level.WARNING)){
```

Comment:
Maybe we should go out from the Warp/Affine reduction and come back to the two step Warp/Affine operation?